### PR TITLE
[NVIDIA XLA:GPU] Introducing training capability for cudnn fused attention (graph rewriter, 3rd out of 3)

### PR DESCRIPTION
**Pull request by Tixxx**
Introducing training support for lowering into cudnn fused attention for the following patterns:
```
for cudnn version >= 8.8.0 and < 8.9.1:
bmm-scale-bias-softmax-dropout-bmm
bmm-scale-bias-softmax-bmm

for cudnn version >=8.9.1:
bmm-scale-bias-softmax-dropout-bmm
bmm-scale-bias-softmax-bmm
bmm-scale-bias-mask-softmax-dropout-bmm
bmm-scale-bias-mask-softmax-dropout-bmm


```
These are commonly used in models like gpt-3 and t5x

This pr also adds a new pass called cudnnfMhaTransposeFusion which fuses adjacent transposes into the fmha call for improved performance.

Note that dropout will use cudnn's dropout node which has a different rng, lowering of such is only enabled using the flag `xla_gpu_fused_attention_use_cudnn_rng`

This pr only contains rewriter changes, and it's the 3rd and final one of the series, links to other PRs are:
stream executor related changes are [here](https://github.com/openxla/xla/pull/3886) (merged)
lhlo lowering and thunk/emitter changes [here](https://github.com/openxla/xla/pull/4019) (merged)

### CODE REVIEW

File: third_party/tsl/tsl/protobuf/dnn.proto [hunk 1]
```diff
@@ -14,11 +14,12 @@ enum DataType {
   kHalf = 2;
   kInt8 = 3;
   kInt32 = 4;
-  kComplexFloat = 5;
-  kComplexDouble = 6;
-  kBF16 = 7;
-  kF8E5M2 = 8;
-  kF8E4M3FN = 9;
+  kInt64 = 5;
```
Comment (cheshire) [inline third_party/tsl/tsl/protobuf/dnn.proto:17]: Ideally, could we avoid changing values of previous versions?

File: workspace2.bzl [hunk 2]
```diff
@@ -32,9 +32,9 @@ def _tf_repositories():
         name = "cudnn_frontend_archive",
         build_file = "//third_party:cudnn_frontend.BUILD",
         patch_file = ["//third_party:cudnn_frontend_header_fix.patch"],
-        sha256 = "bfcf778030831f325cfc13ae5995388cc834fbff2995a297ba580d9ec65ca3b6",
-        strip_prefix = "cudnn-frontend-0.8",
-        urls = tf_mirror_urls("https://github.com/NVIDIA/cudnn-frontend/archive/refs/tags/v0.8.zip"),
+        sha256 = "d8dba9e2607a0c256aa8eacb45b39986ab6f3f24a4d431d4397047a3cb0cd4fb",
```
Comment (cheshire) [inline workspace2.bzl:35]: Could the version bump be moved to another commit?

File: xla/service/gpu/backend_configs.proto [hunk 3]
```diff
@@ -120,18 +120,25 @@ message CudnnfMHABackendConfig {
   // Opaque algorithm number and tuning knobs chosen for this fMHA.
   stream_executor.dnn.AlgorithmProto algorithm = 8;
 
-  // The scaling factor multiplied with the BMM1 result. fmha_scale is 0 if the
+  // The scaling factor multiplied with the BMM1 result. fmha_scale is 1 if the
```
Comment (cheshire) [inline xla/service/gpu/backend_configs.proto:123]: Why the change from 0 to 1?

File: xla/service/gpu/cublas_cudnn.cc [hunk 4]
```diff
@@ -181,7 +254,128 @@ std::string CudnnfMHAKindToString(CudnnfMHAKind kind) {
...
+    case CudnnfMHAKind::kBackwardScaleBiasSoftmaxDropout:
+      return "fmha_bias_softmax_with_dropout_backward";
+    case CudnnfMHAKind::kBackwardScaleBiasSoftmax:
+      return "fmha_bias_softmax_backward";
+  }
+}
+
+// Give fmha instruction a more useful name than "custom-call.42".
+Status SetFMHAInstructionName(HloModule* module, HloInstruction* fmha) {
+  if (fmha->custom_call_target() == kCudnnfMHABmmBmmCallTarget) {
```
Comment (cheshire) [inline xla/service/gpu/cublas_cudnn.cc:281]: Maybe unify all these branches? Like if the call target has a certain prefix, call SetAndUniquifyInstrName on the custom_call_target()?

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 5]
```diff
@@ -59,6 +65,49 @@ bool IsBatchedMatmul(const HloInstruction* instr) {
...
+        }
+      }
+      visit_list.pop();
+    }
+  }
+  return false;
+}
+
+bool IsFirstFwdMatmul(HloInstruction* gemm) {
+  if (!IsBatchedMatmul(gemm) || IsFwdCustomCallTofMHA(*gemm->operand(0)) ||
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:103]: Just return the negation of the value inside the brackets?

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 6]
```diff
@@ -59,6 +65,49 @@ bool IsBatchedMatmul(const HloInstruction* instr) {
...
 }
 
+// We need to check if current gemm is sharing a parent node with a forward
+// fMHA call. We check this by doing a BFS of all operands to see if there's
+// any user that is a forward fMHA custom call.
+bool IsSharingOperandWithFwdMha(HloInstruction* gemm) {
+  for (int64_t i = 0; i < gemm->operands().size(); i++) {
+    std::queue<HloInstruction*> visit_list;
+    visit_list.push(gemm->mutable_operand(i));
+    while (!visit_list.empty()) {
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:75]: Could you add a comment describing the traversal logic a bit more?

File: xla/service/gpu/gpu_fused_mha_runner.h [hunk 7]
```diff
@@ -222,6 +302,107 @@ class FusedMultiHeadedAttentionRunner {
         return std::make_unique<
             se::dnn::LazyOpRunner<se::dnn::FusedMHAScaleBiasMaskSoftmaxOp>>(
             config.algorithm);
+      default:
+        CHECK(false) << "Internal error: unsupported CUDNN MHA kind in FusedMultiHeadedAttentionRunner";
```
Comment (cheshire) [inline xla/service/gpu/gpu_fused_mha_runner.h:306]: Nit: LOG(FATAL)?

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 8]
```diff
@@ -22,13 +22,17 @@ limitations under the License.
 #include <string>
 #include <utility>
```
Comment (cheshire) [inline xla/stream_executor/cuda/cuda_dnn.cc:23]: Could we decouple SE changes in a separate PR?

**Review by cheshire — CHANGES_REQUESTED** [review]

File: xla/service/gpu/backend_configs.proto (hunk 3 above)
Comment (Tixxx) [reply xla/service/gpu/backend_configs.proto:123]: i think 0 was a typo before. We by default set the scale to 1 in cudnn backend runner if no scale node is found in HLO graph.

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 5 above)
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:103]: updated

File: xla/service/gpu/cublas_cudnn.cc (hunk 4 above)
Comment (Tixxx) [reply xla/service/gpu/cublas_cudnn.cc:281]: i will add a switch to return prefix based on call target and call SetAndUniquifyInstrName

File: workspace2.bzl (hunk 2 above)
Comment (Tixxx) [reply workspace2.bzl:35]: yea i moved it to another pr actually https://github.com/openxla/xla/pull/3816 to make sure it passes all CIs before consuming it in this PR

File: third_party/tsl/tsl/protobuf/dnn.proto (hunk 1 above)
Comment (Tixxx) [reply third_party/tsl/tsl/protobuf/dnn.proto:17]: yea updated

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 6 above)
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:85]: added more comments

File: workspace2.bzl (hunk 2 above)
Comment (Tixxx) [reply workspace2.bzl]: the above pr is merged

File: xla/stream_executor/cuda/cuda_dnn.cc (hunk 8 above)
Comment (Tixxx) [reply xla/stream_executor/cuda/cuda_dnn.cc]: I separated the SE changes out to this [pr](https://github.com/openxla/xla/pull/3886). Since the rewriter depends on it, we can review that one first?

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 9]
```diff
@@ -799,8 +1111,29 @@ StatusOr<bool> FuseMultiHeadedAttentionBlock(
...
-      ShapeUtil::MakeTupleShape({output_shape, ShapeUtil::MakeShape(U8, {0})});
+
+  Shape call_shape;
+  // Activation output is used by backward gemm.
+  HloInstruction* activation_output = nullptr;
+
+  std::vector<Shape> output_shapes = {output_shape,
+                                      ShapeUtil::MakeShape(U8, {0})};
+  if (is_training) {
+    // TODO Flush attention will have a different shape in training.
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:1122]: Nit: "flash"?

**Review by cheshire — CHANGES_REQUESTED** [review]
Thanks a lot!

Another question: for evaluation, are there end-to-end tests in JAX which pick up this rewrite? Do we know what limitations does it have, that is, when the rewrite does or does not trigger?

**Comment by Tixxx** [comment]
> Thanks a lot!
> 
> Another question: for evaluation, are there end-to-end tests in JAX which pick up this rewrite? Do we know what limitations does it have, that is, when the rewrite does or does not trigger?

We are planning to add e2e tests in Jax repo to exercise fused mha but haven't created a pull request at the moment. For testing training changes, we have an unit test that calls [this](https://github.com/google-research/t5x/blob/main/t5x/contrib/gpu/t5/layers.py#L51-L123) function in t5x with synthetic input. This is the pattern that we mainly support.
Additional training patterns are supported based on cudnn version which i have listed in the description.
For these cases, the fmha rewriter will trigger.
Other constraints are:

- The qkv sequence lengths must all be <= 512
- the head dimension of qkv tensors must be equal to 64
- Input tensors must be 4d tensors

**Comment by sherhut** [comment]
The SE changes have landed. @Tixxx could you rebase so that we can start reviewing this change?

**Comment by Tixxx** [comment]
> The SE changes have landed. @Tixxx could you rebase so that we can start reviewing this change?

Done.

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 10]
```diff
@@ -49,6 +52,11 @@ auto OptionalConvert(Pattern pattern) {
   return m::AnyOf<HloInstruction>(m::Convert(shared), shared);
 }
 
+template <typename Pattern>
+auto OptionalBitcast(Pattern pattern) {
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:56]: Those tend to be super-expensive to compile, cf. the fix in other mha rewrite PR (for inference)

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 11]
```diff
@@ -5517,9 +5517,13 @@ GetCudnnFusedMHABackwardOperationGraph(
 
 #if (CUDNN_VERSION >= 8901 && TF_ENABLE_CUDNN_FRONTEND)
   // bias backward
-  TF_ASSIGN_OR_RETURN(auto tensor_dbias, CreateCudnnBiasBwdTensor(
-                                             intermediate_ops, p_dims,
-                                             p_strides, dtype, tensor_ds_mask));
+  if (use_bias) {
```
Comment (Tixxx) [inline xla/stream_executor/cuda/cuda_dnn.cc:5520]: @cheshire This was missed from previous PRs and discovered during testing. Let me know if you'd like these changes to be separated.

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 12]
```diff
@@ -358,67 +360,86 @@ StatusOr<bool> IsSupportedBMM2(const HloInstruction* bmm_2,
   return true;
 }
 
-bool MatchDefaultBmmBmm(int64_t bmm2_operand_position, HloInstruction* instr,
-                        HloInstruction** bmm_1, HloInstruction** bmm_2,
-                        std::string& custom_call_name) {
+bool MatchDefaultFwdBmmBmm(int64_t bmm2_operand_position, HloInstruction* instr,
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:339]: Output parameters aren't great, consider returning a struct?

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 13]
```diff
@@ -358,67 +360,86 @@ StatusOr<bool> IsSupportedBMM2(const HloInstruction* bmm_2,
...
     return true;
   }
   return false;
 }
 
 bool MatchSoftmaxDropoutBmm(int64_t bmm2_operand_position,
-                            HloInstruction* instr, HloInstruction** bmm_1,
-                            HloInstruction** bmm_2,
+                            HloInstruction* instr, HloInstruction** bmm_2,
                             HloInstruction** softmax_input,
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:361]: Same nit on output parameters

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 14]
```diff
@@ -515,15 +533,11 @@ bool MatchBmm1ScaleBiasMaskSoftmaxDropoutBmm2(
...
-// fused and unfused cases.
-bool MatchMHAPatternsForCanonicalization(
+bool MatchFwdMHAPatternsForCanonicalization(
     HloInstruction* instr, HloInstruction** bmm_1, HloInstruction** bmm_2,
     HloInstruction** bias, HloInstruction** mask, HloInstruction** scale,
     double& dropout_rate, std::string& custom_call_name,
-    bool& need_canonicalization) {
-  // TJ - TODO add backward support
-
+    bool& need_canonicalization, bool& is_training) {
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:528]: Same nit on output parameters

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 15]
```diff
@@ -558,25 +570,365 @@ bool MatchMHAPatternsForCanonicalization(
...
+
+bool MatchBmm1GradGemm2(
+    HloInstruction* fwd_fmha_call, HloInstruction** bmm_1_grad_2,
+    HloInstruction** bmm_1_grad_1,
+    std::vector<HloInstruction**>& bmms_need_canonicalization) {
+  // bmm1 gradient gemm2 shares the same input as bmm1 gradient gemm1.
+  // Check to see if bmm1 grad gemm1 needs canonicalization or not, if not,
+  // then the shared input is the first operand.
+  int64_t parent_nodex_index =
+      std::find_if(bmms_need_canonicalization.begin(),
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:606]: Nit: absl::c_linear_search?

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 16]
```diff
@@ -558,25 +570,365 @@ bool MatchMHAPatternsForCanonicalization(
...
+      0) {
+    bmms_need_canonicalization.push_back(bmm_2_grad_1);
+  }
+  return true;
+}
+
+bool MatchBmm2GradGemm2(
+    HloInstruction* fwd_fmha_call, HloInstruction** bmm_2_grad_2,
+    bool v_transposed,
+    std::vector<HloInstruction**>& bmms_need_canonicalization) {
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:673]: Nit: absl::Span?

**Review by cheshire — CHANGES_REQUESTED** [review]

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 17]
```diff
@@ -63,6 +72,48 @@ bool IsBatchedMatmul(const HloInstruction* instr) {
...
+// any user that is a forward fMHA custom call.
+bool IsSharingOperandWithFwdMha(HloInstruction* gemm) {
+  for (int64_t i = 0; i < gemm->operands().size(); i++) {
+    std::queue<HloInstruction*> visit_list;
+    visit_list.push(gemm->mutable_operand(i));
+    while (!visit_list.empty()) {
+      HloInstruction* current_instr = visit_list.front();
+      for (int64_t user_index = 0; user_index < current_instr->user_count();
+           user_index++) {
+        HloInstruction* user = current_instr->users()[user_index];
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:86]: Why not just iterate over `current_instr->users()`?

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 18]
```diff
@@ -63,6 +72,48 @@ bool IsBatchedMatmul(const HloInstruction* instr) {
...
+        switch (user->opcode()) {
+          case HloOpcode::kBitcast:
+          case HloOpcode::kReshape:
+          case HloOpcode::kTranspose: {
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:90]: Could you update the function doc to indicate you're looking for a chain of bitcast/reshape/transpose?

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 19]
```diff
@@ -63,6 +72,48 @@ bool IsBatchedMatmul(const HloInstruction* instr) {
   return is_batch_dot;
 }
 
+// We need to check if current gemm is sharing a parent node with a forward
+// fMHA call. We check this by doing a BFS of all operands to see if there's
+// any user that is a forward fMHA custom call.
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:77]: Is it possible to motivate the reader why do we need to find out whether it's shared?

**Review by cheshire — CHANGES_REQUESTED** [review]

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 10 above)
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:93]: i accidentally added std::move to all subpatterns, reverted my changes

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 12 above)
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc]: refactored all output parameters to be returned from struct for both forward and backward calls

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 15 above)
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc]: this is removed

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 16 above)
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc]: this is removed too

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 17 above)
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc]: changed

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 18 above)
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:132]: added doc to reflect this

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 19 above)
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc]: changed doc to explain this

**Comment by Tixxx** [comment]
@cheshire  did a refactoring of both forward and backward function calls to return structs. Please take another look, thanks.

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 20]
```diff
@@ -15,44 +15,90 @@ limitations under the License.
...
+
+  bool has_match = false;
+  std::string matched_custom_call_name;
+};
+
 template <typename Pattern>
 auto OptionalReshape(Pattern pattern) {
   auto shared = m::SharedSubpattern(pattern);
-  return m::AnyOf<HloInstruction>(m::Reshape(shared), shared);
+  return m::AnyOf<HloInstruction>(m::Reshape(pattern), shared);
```
Comment (tdanyluk) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:83]: This should be: 
```
auto shared = m::SharedSubpattern(pattern);
return m::AnyOf<HloInstruction>(m::Reshape(shared), shared);
```

If you write (m::Reshape(pattern), shared); then the compilation can be slow, because the shared pattern is not reused.

Was there a problem with the original code?

Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:83]: ah this was mostly an error from rebase, let me change it back.

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 10 above)
Comment (cheshire) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:93]: Is the compile-time concern resolved?

**Review by cheshire — CHANGES_REQUESTED** [review]

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 10 above)
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:93]: yes, this is kept the same as the other sharedPattern templates

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 21]
```diff
@@ -220,19 +210,21 @@ bool IsComputeCapabilityAndCudnnSupported(
...
+        supported_cudnn_version.minor_version(),
+        supported_cudnn_version.patch());
     return false;
   }
   return true;
 }
 
 bool IsSupportedPrimitiveType(const HloInstruction* bmm) {
-  PrimitiveType dtype = bmm->shape().element_type();
+  auto dtype = bmm->shape().element_type();
```
[the remark this answers is not in the document]
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:268]: yea, reverted.

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 22]
```diff
@@ -151,65 +195,11 @@ double GetDropoutRateFromHlo(HloInstruction* dropout) {
...
-      "Found invalid FMHA custom-call target while setting custom-call name");
-}
-
 bool IsComputeCapabilityAndCudnnSupported(
     stream_executor::CudaComputeCapability cc,
     stream_executor::dnn::VersionInfo cudnn_version,
-    stream_executor::StreamExecutor* stream_exec) {
-  // return true;
+    stream_executor::StreamExecutor* stream_exec,
+    stream_executor::dnn::VersionInfo supported_cudnn_version) {
```
[the remark this answers is not in the document]
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:243]: supported_cudnn_version was added since we have different training features available in different versions of cudnn. for version 8.8.0 - 8.9.1, cudnn only supports additive bias and mask. For 8.9.1+ cudnn supports any type of mask. We use this utility function to decide what we would rewrite based on cudnn version.
The previous
`// return true;`
was left there by mistake though..

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 23]
```diff
@@ -99,25 +142,26 @@ bool IsScaledMaskedFusedSoftmaxCall(const HloInstruction* instr) {
 // |   /
 // root
 auto GetUnfusedReduceMaxSumSoftmaxPattern(
-    HloInstruction** softmax_input = nullptr) {
```
[the remark this answers is not in the document]
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:102]: This is actually a function signature change which is necessary for training cases, we need to match for additional nodes in softmax.

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 24]
```diff
@@ -63,6 +72,48 @@ bool IsBatchedMatmul(const HloInstruction* instr) {
...
+          default:
+            break;
+        }
+      }
+      visit_list.pop();
+    }
+  }
+  return false;
+}
+bool IsFirstFwdMatmul(HloInstruction* gemm) {
```
[the remark this answers is not in the document]
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:152]: added

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 25]
```diff
@@ -358,67 +336,86 @@ StatusOr<bool> IsSupportedBMM2(const HloInstruction* bmm_2,
...
-              bmm2_operand_position,
-              m::Op(bmm_1).WithPredicate(IsBatchedMatmul).WithOneUse());
+          .WithOperand(bmm2_operand_position,
+                       m::Op(bmm_1).WithPredicate(IsBatchedMatmul));
 
-  if (Match(instr, default_bmm_bmm_pattern)) {
+  // If any of bmm1's operands is coming from a forward fMHA call, then return
+  // false
+  if (Match(instr, default_bmm_bmm_pattern) && IsFirstFwdMatmul((*bmm_1))) {
+    is_training = (*bmm_1)->user_count() == 2;
```
[the remark this answers is not in the document]
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc]: added some explanation to this

**Comment by ddunl** [comment]
Both reviewers are currently on vacation! I'm trying to find someone else to take a look. Sorry about the delay!

**Comment by Tixxx** [comment]
> Both reviewers are currently on vacation! I'm trying to find someone else to take a look. Sorry about the delay!

Thanks David.
I just rebased to resolve conflicts.

**Review by tdanyluk — APPROVED** [review]
I'm approving this, because it looks thoroughly tested plus cheshire@ has reviewed it earlier and most of his comments has been resolved.
Thank you for the huge contribution.

**Comment by Tixxx** [comment]
> I'm approving this, because it looks thoroughly tested plus cheshire@ has reviewed it earlier and most of his comments has been resolved. Thank you for the huge contribution.

Thanks so much for the review and approval!

**Review by reedwm — APPROVED** [review]
Blindly approving since it's late in @tdanyluk's timezone and this needs reapproval since this was rebased onto main.

**Comment by Tixxx** [comment]
> Blindly approving since it's late in @tdanyluk's timezone and this needs reapproval since this was rebased onto main.

ah ok, didnt know it would need a reapproval after rebase. Thanks for the quick turn-around @reedwm

**Comment by Tixxx** [comment]
hi @ddunl  do you see any issues from internal CI? Please let me know so i can address them asap.

**Comment by ddunl** [comment]
I know @reedwm is looking

**Comment by reedwm** [comment]
I am working on merging this, I'll let you know if there are any major issues
